pennylane/transforms/optimization/optimization_utils.py

```diff
@@ -47,146 +41,105 @@ def find_next_gate(wires, op_list):
     return next_gate_idx
 
 
-def _zyz_to_quat(angles):
-    """Converts a set of Euler angles in ZYZ format to a quaternion."""
-    qw = cos(angles[1] / 2) * cos(0.5 * (angles[0] + angles[2]))
-    qx = -sin(angles[1] / 2) * sin(0.5 * (angles[0] - angles[2]))
-    qy = sin(angles[1] / 2) * cos(0.5 * (angles[0] - angles[2]))
-    qz = cos(angles[1] / 2) * sin(0.5 * (angles[0] + angles[2]))
-
-    return stack([qw, qx, qy, qz])
-
-
-def _quaternion_product(q1, q2):
-    """Compute the product of two quaternions, q = q1 * q2."""
-    qw = q1[0] * q2[0] - q1[1] * q2[1] - q1[2] * q2[2] - q1[3] * q2[3]
-    qx = q1[0] * q2[1] + q1[1] * q2[0] + q1[2] * q2[3] - q1[3] * q2[2]
-    qy = q1[0] * q2[2] - q1[1] * q2[3] + q1[2] * q2[0] + q1[3] * q2[1]
-    qz = q1[0] * q2[3] + q1[1] * q2[2] - q1[2] * q2[1] + q1[3] * q2[0]
-
-    return stack([qw, qx, qy, qz])
-
-
-def _singular_quat_to_zyz(qw, qx, qy, qz, y_arg, abstract_jax=False):
-    """Compute the ZYZ angles for the singular case of qx = qy = 0"""
-    # pylint: disable=too-many-arguments
-    z1_arg1 = 2 * (qx * qy + qz * qw)
-    z1_arg2 = 1 - 2 * (qx**2 + qz**2)
+def _try_no_fuse(angles_1, angles_2):
+    """Try to combine rotation angles without trigonometric identities
+    if some angles in the input angles vanish."""
+    angles_1 = qml.math.asarray(angles_1)
+    angles_2 = qml.math.asarray(angles_2)
+    # This sum is only computed to obtain a dtype-coerced object that respects
+    # TensorFlow's coercion rules between Python/NumPy objects and TF objects.
+    _sum = angles_1 + angles_2
+    # moveaxis required for batched inputs
+    phi1, theta1, omega1 = qml.math.moveaxis(qml.math.cast_like(angles_1, _sum), -1, 0)
+    phi2, theta2, omega2 = qml.math.moveaxis(qml.math.cast_like(angles_2, _sum), -1, 0)
+
+    if qml.math.allclose(omega1 + phi2, 0.0):
```

Comment:
Should the following logic also take care of modulo-pi-ness (i.e. special cases when angles are close to +/- N * pi), or is it taken care of somewhere else or not applicable?

Reply:
That's a fair question! The previous version of this function did not have any guarantees for the output range, do we want to introduce this guarantee? :thinking: I'd hope that it is not a problem if all ML interfaces define the derivative of `mod` everywhere :grimacing:

Reply:
I have nothing specific in mind, I'm just wondering if there are other special cases away from zero which we could handle here. For instance, I think `omega`, `phi` are taken in [-pi, pi]. If they both are almost pi, should we not merge them and also return `qml.math.stack([phi1, theta1 + theta2, omega2])`? Not a blocker from me in any case.

Reply:
Ah, I understand! Sorry, I thought you were thinking of doing `np.mod(theta1+theta2, 2*np.pi)` or something like that.

We could, in principle, do the following:
```python
if qml.math.allclose(qml.math.mod(omega1 + phi2+0.1, 2* np.pi)-0.1, 0.0):
    return qml.math.stack([phi1 + omega1 + phi2, theta1 + theta2, omega2])
```
If they are zero, taken together, this doesn't change things, and if they are other multiples of 2*pi, they lead to either a global phase or the identity, which can be pulled up front.

Similarly, we could think about pulling factors of pi in the `RY` part through.

I'm happy to include these things, but I'm not sure how much optimization we want :)

CC @trbromley do you have an opinion here?

Reply:
> I thought you were thinking of doing `np.mod(theta1+theta2, 2*np.pi)` or something like that.

That too :D it was a bit left to the imagination. But first and foremost, I was thinking introducing this kind of logic `qml.math.allclose(qml.math.mod(omega1 + phi2+0.1, 2* np.pi)-0.1, 0.0)`. Let's see what @trbromley thinks.

Reply:
Overall, I think our broad preference is towards:
- Code correctness, i.e., incorrect numeric answers are not provided.
- Avoiding over-optimized simplifications during higher-level Python passes. Going forward (outside the scope of this PR), we can look to add low-level compiler passes that could work across the ecosystem (Catalyst and Python PennyLane). Not sure exactly what this looks like right now, it's perhaps linked to how we decide to get transforms and program capture to work together.

For this PR, I'd recommend we focus on code correctness, but leave it up to you for the particular decision here.

Reply:
In this case I agree to go for correctness and not introduce this change here.